### PR TITLE
[diag] support send security processed frames

### PR DESCRIPTION
**Pull request by bukepo**
This commit extends the `diag frame` command by adding an argument `-s` to indicate whether the frame has been encrypted or not, so that the diag commands can be used to test frames having security processed in the host.

### CODE REVIEW

File: src/core/diags/factory_diags.cpp [hunk 1]
```diff
@@ -221,7 +221,24 @@ Error Diags::ProcessFrame(uint8_t aArgsLength, char *aArgs[])
     Error    error = kErrorNone;
     uint16_t size  = OT_RADIO_FRAME_MAX_SIZE;
 
-    VerifyOrExit(aArgsLength == 1, error = kErrorInvalidArgs);
+    memset(&mTxPacket->mInfo, 0, sizeof(mTxPacket->mInfo));
+
+    while (aArgsLength > 1)
+    {
+        if (!strcmp(aArgs[0], "-s"))
```
Comment (abtink) [inline src/core/diags/factory_diags.cpp:228]: Use `StringMatch`?
```suggestion
        if (StringMatch(aArgs[0], "-s"))
```

File: src/core/diags/factory_diags.cpp [hunk 2]
```diff
@@ -221,7 +221,24 @@ Error Diags::ProcessFrame(uint8_t aArgsLength, char *aArgs[])
     Error    error = kErrorNone;
     uint16_t size  = OT_RADIO_FRAME_MAX_SIZE;
 
-    VerifyOrExit(aArgsLength == 1, error = kErrorInvalidArgs);
+    memset(&mTxPacket->mInfo, 0, sizeof(mTxPacket->mInfo));
```
Comment (abtink) [inline src/core/diags/factory_diags.cpp:224]: Can use `ClearAllBytes()`?
```suggestion
    ClearAllBytes(mTxPacket->mInfo);
```

File: src/core/diags/factory_diags.cpp [hunk 3]
```diff
@@ -221,7 +221,24 @@ Error Diags::ProcessFrame(uint8_t aArgsLength, char *aArgs[])
...
+
+    while (aArgsLength > 1)
```
Comment (abtink) [inline src/core/diags/factory_diags.cpp:226]: Do we need `while()` here?
We can allow a single `-s` option. Any benefit to allow multiple

File: src/core/diags/factory_diags.cpp [hunk 4]
```diff
@@ -468,6 +485,7 @@ void Diags::TransmitPacket(void)
 
     if (!mIsTxPacketSet)
     {
+        memset(&mTxPacket->mInfo, 0, sizeof(mTxPacket->mInfo));
```
Comment (abtink) [inline src/core/diags/factory_diags.cpp:488]:
```suggestion
        ClearAllBytes(mTxPacket->mInfo);
```

File: src/core/diags/factory_diags.cpp [hunk 5]
```diff
@@ -221,7 +221,24 @@ Error Diags::ProcessFrame(uint8_t aArgsLength, char *aArgs[])
...
+        else
+        {
+            break;
+        }
+
+        aArgs++;
+        aArgsLength--;
+    }
+
+    VerifyOrExit(aArgsLength >= 1, error = kErrorInvalidArgs);
```
Comment (abtink) [inline src/core/diags/factory_diags.cpp:241]: We seem to expect and allow one arg only?
```suggestion
    VerifyOrExit(aArgsLength == 1, error = kErrorInvalidArgs);
```

**Review by abtink — COMMENTED** [review]
Looks good.  Thanks. 
Some smaller suggestions below:

File: src/core/diags/factory_diags.cpp (hunk 3 above)
Comment (bukepo) [reply src/core/diags/factory_diags.cpp:226]: I'm going to add flags for other fields in txinfo like csma/ca. Do you think I should add `while` in that PR?

Comment (abtink) [reply src/core/diags/factory_diags.cpp:226]: Yes please. Make sense to add it then.

File: src/core/diags/factory_diags.cpp [hunk 6]
```diff
@@ -216,11 +217,38 @@ Diags::Diags(Instance &aInstance)
...
+    mTxPacket->mInfo.mTxInfo.mCslPresent           = false;
+    mTxPacket->mInfo.mTxInfo.mIsSecurityProcessed  = false;
+}
+
 Error Diags::ProcessFrame(uint8_t aArgsLength, char *aArgs[])
 {
     Error    error = kErrorNone;
     uint16_t size  = OT_RADIO_FRAME_MAX_SIZE;
 
+    ResetTxPacket();
```
Comment (zhanglongxia) [inline src/core/diags/factory_diags.cpp:240]: Should we reset the frame after all parameters are passed the verification?

File: src/core/diags/factory_diags.cpp [hunk 7]
```diff
@@ -216,11 +217,38 @@ Diags::Diags(Instance &aInstance)
...
+
+    if (aArgsLength >= 1)
+    {
+        if (StringMatch(aArgs[0], "-s"))
+        {
+            mTxPacket->mInfo.mTxInfo.mIsSecurityProcessed = true;
```
[the remark this answers is not in the document]
Comment (bukepo) [reply src/core/diags/factory_diags.cpp:246]: It's reset in TransmitPacket().

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/core/diags/README.md [hunk 8]
```diff
@@ -77,10 +78,14 @@ Stop transmitting continuous carrier wave.
...
 ```
 
-### diag frame \<frame\>
+### diag frame
+
+Usage: `diag frame [-s] <frame>`
 
 Set the frame (hex encoded) to be used by `diag send` and `diag repeat`. The frame may be overwritten by `diag send` and `diag repeat`.
 
+Specify `-s` to skip security process in radio layer.
```
Comment (abtink) [inline src/core/diags/README.md:87]:
```suggestion
Specify `-s` to skip security processing in radio layer.
```

**Review by abtink — APPROVED** [review]
Looks great. Thanks. 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**